boards/arm/efr32mg_sltb004a/board.h

```diff
@@ -13,4 +13,10 @@
 #define CCS811_PWR_ENABLE_GPIO_PIN	14
 #endif /* CONFIG_CCS811 */
 
+/* SI7021 specific pins */
```

Comment:
There should be a binding file for the sensor, and the `vin-supply` property from the base `power.yaml` used to identify this signal.  Then the driver should be responsible for enabling the device power on initialization.

Reply:
Found the docs, and correction: `supply-gpios` property.  See https://docs.zephyrproject.org/latest/reference/peripherals/regulators.html?highlight=regulators#regulators

Reply:
OK, I'm surprised the existing si7006 driver works with this sensor.

In that case we just need better documentation on which compatibles work with the sensor, and which driver to select for a given sensor.  `silabs,7021` should select the si7006 driver, and we should be able to find that fact from #30104.

So just integrating `supply-gpios` support into that driver could be the right path forward.

Reply:
I was actually thinking that defining the enable pin in the dts is nicer, but I was unaware of the regulator subsystem, thanks for that input! My initial guess was, that since it is something specific to the board, and not the device itself, it should not go into the sensor driver. But I see what you mean! 

So I will go ahead, include the enable mechanism in the si7006 driver, and set the correct pin in the board overlay for the example. 

I'm not 100% sure what you mean with "silabs,7021 should select si7006 driver"? So as far as I understand, I just use "silabs,si7006" as compatible property in the device tree overlay, right? Or should I still create a binding for silabs,si7021 as well?

Reply:
> I'm not 100% sure what you mean with "silabs,7021 should select si7006 driver"? So as far as I understand, I just use "silabs,si7006" as compatible property in the device tree overlay, right?

The devicetree should use:

    compatible = "silabs,si7021", "silabs,si7006";

which specifies the hardware from most specific to least specific.  The si7006 driver would recognize the node, but so would an si7021 driver if one were present.

Reply:
If I remove the specific si7021 driver and use 

`compatible = "silabs,si7021", "silabs,si7006";`

instead, how should the config look then? Should there be a CONFIG_SI7021 which selects CONFIG_SI7006? Or do we expect the user to know that the driver for the si7006 can be used with this device as well?

Reply:
See #30104.  The driver would remain selected by `CONFIG_SI7006`.  The user has to select the driver that implements support for the compatible.

Reply:
I have changed the compatible property to include "silabs,si7006", that works quite well. 

I'm however a bit unsure as to how to proceed with the power enabling for that sensor. The sample application that I included here I created specifically to showcase the current problem with the I2C multiplexing as described above. 

The si7021 on the Thunderboard 2 Sense shares the same power enable transistor with 2 other sensors. I don't know how much sense it makes to include the supply-gpios mechanism in the driver then, especially since it has nothing specifically to do with that sensor. As far as I understand, the correct way would be to include a dedicated regulator node in the dts of the board itself, right? If I do so, should the regulator_enable(...) call happen inside the board init procedure, or should that be left to the application developer (as they might not even want to use that sensor)? 

One possibility would be to include the regulator in the sample device tree overlay, and maybe put a note in the README. 

Thanks a lot for your continuous input on this!

Reply:
Ah; yes, I have a couple of those, and they informed the design of the regulator API.  I hadn't realized that one of the rails had multiple sensors on it.

The "correct" solution is to add a fixed regulator for the `ENV_SENSE_ENABLE` signal in the `efr32mg_sltb004a` devicetree, and add a `vin-supply` reference to it to each of the sensors.  Then in the init routine for each sensor use [regulator_enable()](https://docs.zephyrproject.org/latest/reference/peripherals/regulators.html#c.regulator_enable) to ensure power is provided while the sensor is being used.

Do you feel up to giving that a shot?  If you'd rather not, I can do it: I do want to have more examples of how the regulator API is supposed to work (and to confirm that it meets user needs).

Reply:
I'm a bit short on time right now, but I would really like to try giving it a shot. I will most likely still need input, since I haven't found any other sensor driver using the regulator subsystem so far. 

I'm also still a bit uncertain about how the whole i2c configuration is supposed to be handled nicely. It seems from #9042, that originally the i2c location was defined so that one could use the three environmental sensors on the Thunderboard 2. When support for the ccs811 sensor was added in #25376, this was changed, so that the i2c1 peripheral is now routed to the pins connecting to that sensor. As pointed out in that pr, one cannot (easily) use both the ccs811 and the three environment sensors at the same time. Since the i2c1 pin location setting is part of the dts of the board, I would still like to include the sample application, to show how to set the board overlay correctly so that the environment sensor group can be used. Do you have any input on that? 

I also think that #30104 is really important. It is quite hard to figure out which drivers work with which sensors.

Reply:
Also out of curiosity, the `supply-gpios` property is of type `phandle`. So do you still need to configure a regulator node in the dts for that? Otherwise, shouldn't it be a `phandle-array`?

Reply:
> I'm a bit short on time right now, but I would really like to try giving it a shot. I will most likely still need input, since I haven't found any other sensor driver using the regulator subsystem so far.

Yeah, see, that's the problem.  I'm the one who designed the regulator system, and I could answer my own questions faster.  It also ties to device power management which is something I'm responsible for.

Well, ping me when you're about to start, and I'll ping you when I'm about to start, and the one who gets there first gets to take it on.

> I'm also still a bit uncertain about how the whole i2c configuration is supposed to be handled nicely.

I think that a complete answer involves using the pinmux API that's not getting much love right now, though it would be possible to be more flexible in selecting which static configuration each I2C peripheral should use.

Reply:
> Also out of curiosity, the `supply-gpios` property is of type `phandle`.  So do you still need to configure a regulator node in the dts for that? Otherwise, shouldn't it be a `phandle-array`?

No, it's `phandle` because you can only have one, and it can't contain non-phandle integers (phandle-array is a list of interleaved phandles and integers).

And no, if you use `supply-gpios` it's not a regulator it's just a pin, so it doesn't support reference counting.  If it was a dedicated power supply (like for ccs811 on this board) it'd be `supply-gpios` completely under control of the ccs811 driver.  Since it's shared it has to be a regulator described by `vin-supply` that all three drivers use and that will keep the power flowing as long as at least one driver is using it.

Reply:
>> I'm also still a bit uncertain about how the whole i2c configuration is supposed to be handled nicely.

>  think that a complete answer involves using the pinmux API that's not getting much love right now, though it would be possible to be more flexible in selecting which static configuration each I2C peripheral should use.

Just as a hint: The EFM/EFR32 does not have a global pinmux peripheral like most of the other SoCs. In each peripheral (like I2C) there are dedicated location registers, which control which peripheral signal is routed to which pin.
There is no pinmux driver in zephyr for EFM32/EFR32. Every peripheral driver handles its own location registers.

So to solve this problem here, it should be fine to solve this within the i2c driver.

Reply:
@chrta I think that the I2C driver (or perhaps the driver for the sensor that needs an I2C controller to talk to it) should still be doing that configuration with a standard pinmux API.  Though I don't know what's happening with pinmux and whether that use case is being taken into account.  FYI @carlescufi

Reply:
> > Also out of curiosity, the `supply-gpios` property is of type `phandle`.  So do you still need to configure a regulator node in the dts for that? Otherwise, shouldn't it be a `phandle-array`?
> 
> ~No, it's `phandle` because you can only have one, and it can't contain non-phandle integers (`phandle-array` is a list of interleaved phandles and integers).~ wait, no, I think that's completely wrong. I'm going to have to look into that.
> 
> And no, if you use `supply-gpios` it's not a regulator it's just a pin, so it doesn't support reference counting. If it was a dedicated power supply (like for ccs811 on this board) it'd be `supply-gpios` completely under control of the ccs811 driver. Since it's shared it has to be a regulator described by `vin-supply` that all three drivers use and that will keep the power flowing as long as at least one driver is using it.

I might have done something wrong, I just tried defining `supply-gpios = <&gpiof 9 GPIO_ACTIVE_HIGH>; ` in the si7021 dts node, and the dt compiler complained about it not following the given `supply-gpios = < &foo >` structure. 
Have I misunderstood the syntax for that?

Reply:
No, you were correct.  I'm fixing that.

Reply:
If you want to proceed in parallel, make it `phandle-array`.  I'm also looking into the way the board init works.  I think we can do much better in supporting these sensors and managing their power configuration.